### PR TITLE
Training browser

### PR DESCRIPTION
**Pull request by a3ahmad**
## Types of changes

Adds a simple browser to chart training results from multiple runs

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Motivation and Context / Related issue

Adds a quick and easy way to browse/compare results

## How Has This Been Tested (if it applies)

I ran a few different training runs, with different algorithms and use this to compare them

## Checklist

- [ ] The documentation is up-to-date with the changes I made.
- [X] I have read the [**CONTRIBUTING**](../../CONTRIBUTING.md) document and completed the CLA (see **CONTRIBUTING**).
- [ ] All tests passed, and additional code has been covered with new tests.

### CODE REVIEW

**Comment by luisenp** [comment]
@a3ahmad This looks potentially pretty nice! Would you mind adding some brief instruction on how to run so that I can play around with it before making any suggestions?

Also, probably the best place to put this  would be in the `diagnostics` folder (we might change later this folder to something like "visualizations").

**Comment by a3ahmad** [comment]
Sure, just run it with 'python training_browser.py' (you will need PyQt5 and PyQtChart installed -- pip and conda provide it)

You'll see a full size window with a (dockable) pane on the left listing all the available experiments in a table. Click on each to chart their results (or shift-click to select/overlay multiple results).  The toolbar currently has check-boxes to use a logarithmic axis, and to filter out outliers (I don't like the current implementation as it removes the points, rather than just rescale the Y axis)

**Comment by a3ahmad** [comment]
Ok, I've moved it to diagnostics.  So now you can run it with:

python -m mbrl.diagnostics.training_browser

I run it from the root mbrl_lib folder and it will infer current_working_directory/exp. Alternatively you can run:

python -m mbrl.diagnostics.training_browser --experiments [path to the experiment folder]

**Comment by luisenp** [comment]
This is awesome @a3ahmad, thanks a lot! A couple of suggestions:

- Rather than reading the experiment description from the path, it would be better to read the yaml config generated by Hydra, which is saved in `.hydra/config.yaml`, under the same folder that has `results.csv`. There we can get some of the same information (env, algorithm, experiment_name) except the date/time; we can also get other info like the seed used. This is to avoid users that have a different folder structure to get weird info in the browser. For example, when you run an algorithm with `--multirun` in Hydra, you'll get an extra folder with an integer index, and in this case the current browser code crashes. I could potentially add some date info in a metadata file (or maybe save it in the config) to keep it available for the browser, if you think that's useful. 
- Can we add another tab to the browser that plots the mean and standard deviation plot over the different results, as shown below? We can probably pre-compute the data frame created in line line 117, and the serve depending on the index row clicked. This makes sense for things that have the same experiment_name+env+algorithm config+datetime though.

image

Would you be OK with adding this changes?

**Comment by luisenp** [comment]
BTW, `black` was updated recently which broke the CI tests. I've already updated `master`, so you can just rebase and force push here.

**Comment by a3ahmad** [comment]
Those are great suggestions, I’ll get on them

**Comment by a3ahmad** [comment]
@luisenp, what kind of rolling window do you use for the mean/std-dev plots?

**Comment by luisenp** [comment]
> @luisenp, what kind of rolling window do you use for the mean/std-dev plots?

@a3ahmad For the plot like the one I showed I usually just plot with no smoothing. Just average/std over multiple runs (in the case of that plot it was 10 seeds), for all time steps. We could add the window as an optional argument, seems like a useful feature.

**Comment by a3ahmad** [comment]
Ohh, I got you. I need to experiment with multiple runs!

**Comment by luisenp** [comment]
Hi @a3ahmad, thanks again for working on this! I noticed you added a checkbox labeled "Display As Distribution", but currently appears greyed out for me. Is there anything I need to do to enable it?

**Comment by a3ahmad** [comment]
Oh! You need you multiselect several runs of the same length.

The graph doesn’t look right to me, though. I need to investigate.

**Comment by a3ahmad** [comment]
Updated to swap from QtCharts to matplotlib. Simplifies code and works better.

File: mbrl/diagnostics/training_browser.py [hunk 1]
```diff
@@ -172,12 +173,13 @@ def onExperimentsSelectionChanged(self, sel1=None, sel2=None):
...
                 var_series = df.std(axis=1)
                 
-                mean_line = self.axes.plot(time_series, mean_series)
+                mean_line = self.axes.plot(time_series, mean_series, label=self.experiment_results[rowIndex])
                 self.axes.fill_between(time_series, mean_series + var_series, mean_series - var_series, color = mean_line[0].get_color(), alpha = 0.25)
 
             if self.logYAxisCheckbox.checkState():
                 self.axes.semilogy()
 
+            self.axes.legend()
```
Comment (luisenp) [inline mbrl/diagnostics/training_browser.py:182]: This is looking great! One comment and I'm not sure if this is awkward to add. Is there any way we can have the user choose labels for the legend of the aggregated plot, and same for xlabel and ylabel?

File: mbrl/diagnostics/training_browser.py [hunk 2]
```diff
@@ -172,12 +173,13 @@ def onExperimentsSelectionChanged(self, sel1=None, sel2=None):
                 mean_series = df.mean(axis=1)
                 var_series = df.std(axis=1)
                 
-                mean_line = self.axes.plot(time_series, mean_series)
+                mean_line = self.axes.plot(time_series, mean_series, label=self.experiment_results[rowIndex])
```
Comment (luisenp) [inline mbrl/diagnostics/training_browser.py:176]: Is there any way to have a button that has the option to save the plot to some folder? Even something as simple as a showing a text box for the user to write the path and use `matplotlib` to save it somewhere.

File: mbrl/diagnostics/training_browser.py [hunk 3]
```diff
@@ -16,6 +16,7 @@
 from matplotlib.figure import Figure
```
Comment (luisenp) [inline mbrl/diagnostics/training_browser.py:16]: One more thing, if you haven't already, can you install https://pre-commit.com/ to run the linter, mypy, etc. automatically?

**Review by luisenp — CHANGES_REQUESTED** [review]
This is great, thanks a lot! Left a few minor comments, but this is almost ready from my point of view.

**Pull request merged**